### PR TITLE
(QA-2444) Update API to use block syntax

### PR DESCRIPTION
**Pull request by zreichert**

### CODE REVIEW

File: lib/rototiller/task/rototiller_task.rb [hunk 1]
```diff
@@ -35,18 +37,32 @@ def self.define_task(*args, &task_block)
...
+      # @option args [String] :name The environment variable
+      # @option args [String] :default The default value for the environment variable
+      # @option args [String] :message A message describing the use of this variable
+      #
+      # for block {|a| ... }
+      # @yield [a] Optional block syntax allows you to specify information about the environment variable, available methods track hash keys
+      def add_env(*args,&block)
+        raise ArgumentError.new("add_env takes a block or a hash") if !args.empty? && block_given?
+        attributes = [:name, :default, :message]
+        add_param(@env_vars, EnvVar, attributes, args, &block)
```
Comment (er0ck) [inline lib/rototiller/task/rototiller_task.rb:50]: so cool.  dig it

File: lib/rototiller/task/rototiller_task.rb [hunk 2]
```diff
@@ -93,6 +109,30 @@ def define(args, &task_block)
       def set_verbose(verbosity=true)
         @verbose = verbosity
       end
+
+      def add_param(collection, param_class, param_array, args, &block)
+
+        if block_given?
+
+          block_syntax_obj = Object.new
+          block_syntax_obj.class.class_eval { param_array.each { |i| attr_accessor i } }
```
Comment (er0ck) [inline lib/rototiller/task/rototiller_task.rb:118]: :cold_sweat: if it works, it works?

File: lib/rototiller/utilities/env_var.rb [hunk 3]
```diff
@@ -13,13 +13,15 @@ class EnvVar
...
-  def initialize(var, default=nil, message)
-    @var = var
-    @message = message
-    @default = default
+  # @param [Hash] attribute_hash hash of information about the environment variable
+  # @option attribute_hash [String] :name The environment variable
+  # @option attribute_hash [String] :default The default value for the environment variable
+  # @option attribute_hash [String] :message A message describing the use of this variable
+  def initialize(attribute_hash)
+    rasie(ArgumentError, 'A name must be supplied to add_env') unless attribute_hash[:name]
```
Comment (er0ck) [inline lib/rototiller/utilities/env_var.rb]: raise.

i guess we need a spec test for when `!attribute_hash[:name]`

Comment (er0ck) [inline lib/rototiller/utilities/env_var.rb]: @ztr, please add a spec test on this

**Comment by er0ck** [comment]
@zreichert the move of rakefile_tools is a good one (whoops) but will not work on its own.  please contain the changes to acceptance to find the library code in the same commit.

edit: oh. maybe it will... depending on how we set beaker's load_path

File: acceptance/tests/rototillerTask/command_flags.rb [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    hash = h.dup
+    [:block_syntax].each do |key|
+      hash.delete(key)
+    end
+    return hash
+  end
+
+  command_flags = [
+      {:name => '--hash-syntax',  :value => 'wow such hash syntax'},
+      {:name => '--us-a-block',   :value => 'wow much block syntax', :block_syntax => true}
```
Comment (er0ck) [inline acceptance/tests/rototillerTask/command_flags.rb]: typo `--us`?

File: acceptance/tests/rototillerTask/command_flags.rb [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+
+  @task_name    = 'command_flag_testing_key_value'
+  rakefile_contents = <<-EOS
+$LOAD_PATH.unshift('/root/rototiller/lib')
+require 'rototiller'
+
+Rototiller::Task::RototillerTask.define_task :#{@task_name} do |t|
+    #{create_rakefile_task_segment(command_flags)}
+    t.command = 'echo'
```
Comment (er0ck) [inline acceptance/tests/rototillerTask/command_flags.rb]: we probably want to use a command that actually takes flags here.  
i also think we should be testing with multiple flags.  lots of existing Rakefiles don't split(' ') on the command args and beaker/others complain about the flags because they are malformed somehow.

File: acceptance/tests/rototillerTask/command_flags.rb [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        command_regex = /#{flag[:name]} #{flag[:value]}/
+        rototiller_output_regex = /The CLI flag #{flag[:name]} will be used with value #{flag[:value]}/
+        assert_match(command_regex, result.stdout, "The expected output from rototiller was not observed")
+        assert_match(rototiller_output_regex, result.stdout, "The flag #{flag[:name]} was not observed on the command line")
+      end
+    end
+  end
+end
+
+
```
Comment (er0ck) [inline acceptance/tests/rototillerTask/command_flags.rb]: nit: spurious EOF lines

File: acceptance/tests/rototillerTask/environment_variables.rb [hunk 7]
```diff
@@ -9,23 +9,46 @@ def create_rakefile_task_segment(envs)
...
+      {:name => 'DEFAULT-EXISTS',       :message => 'default, previously exists',
+       :default => 'present default value',   :exists => true},
+      {:name => 'DEFAULT-NO_EXISTS',    :message => 'default, does notpreviously exist',
+       :default => 'DEFAULT-NO_EXISTS: notpresent default value', :exists => false},
+      {:name => 'NO_DEFAULT-EXISTS-BLOCK',    :message => 'no default, previously exists',
+       :exists => true, :block_syntax => true},
+      {:name => 'DEFAULT-EXISTS-BLOCK',       :message => 'default, previously exists',
+       :default => 'present default value',   :exists => true, :block_syntax => true},
+      {:name => 'DEFAULT-NO_EXISTS-BLOCK',    :message => 'default, does notpreviously exist',
+       :default => 'DEFAULT-NO_EXISTS-BLOCK: notpresent default value', :exists => false,
```
Comment (er0ck) [inline acceptance/tests/rototillerTask/environment_variables.rb]: some inconsistency in the formatting here.  i suspect it's largely because DEFAULT-EXISTS doesn't have its name in default.  but maybe that was by design, i don't recall

File: lib/rototiller/task/rototiller_task.rb [hunk 8]
```diff
@@ -2,13 +2,15 @@
...
 require 'rototiller/utilities/flag_collection'
+require 'rototiller/utilities/command_flag'
 require 'rake/tasklib'
 
 module Rototiller
   module Task
     class RototillerTask < ::Rake::TaskLib
       attr_accessor :name
       attr_accessor :command
+      attr_accessor :command_arg
```
Comment (er0ck) [inline lib/rototiller/task/rototiller_task.rb]: why is this needed?

File: lib/rototiller/utilities/command_flag.rb [hunk 9]
```diff
@@ -32,4 +34,21 @@ def describe_flag_state
...
+    if !h[:name] && !h[:value]
+      error = name_error << "\n" << value_error
+    elsif h[:name] && !h[:value]
+      error = value_error
+    elsif !h[:name] && h[:value]
+      error = name_error
+    end
+
+    raise(ArgumentError, error) if error
+  end
```
Comment (er0ck) [inline lib/rototiller/utilities/command_flag.rb:53]: hmmm.  we may want to consider the creation of a validator class for hash/block receiving methods.
we already kinda have a schema, somewhat defined in add_env() add_flag() in the attributes argument.
this code probably should have been replicated into env_var (thus the value of a separate validator or validator adapter)

Comment (zreichert) [inline lib/rototiller/utilities/command_flag.rb:53]: @er0ck I agree.  The values are in flux at this point :value ~ :default.  Should this work be ticketed?

File: spec/rototiller/utilities/env_var_spec.rb [hunk 10]
```diff
@@ -74,4 +78,9 @@
       end
     end
   end
+
+  it 'errors when no name is provided' do
+    no_name = {:default => 'default value', :message => 'This is the message'}
+    expect{EnvVar.new(no_name)}.to raise_error(ArgumentError, 'A name must be supplied to add_env')
+  end
```
Comment (er0ck) [inline spec/rototiller/utilities/env_var_spec.rb:85]: :clap:

File: lib/rototiller/utilities/command_flag.rb (hunk 9 above)
Comment (er0ck) [inline lib/rototiller/utilities/command_flag.rb:53]: yeah, later work here is fine.

**Comment by er0ck** [comment]
acceptance tests work and pass.  nice
